man/krusgen.1
.TH krusgen 1 "23 April 2017" "v1.3"
.SH NAME
.PP
\fBkrusgen\fP \- command line maze generator.
.SH SYNOPSIS
.PP
\fBkrusgen\fP [\fB\fCOPTIONS\fR]
.SH DESCRIPTION
.PP
Generates a maze of given size, using Kruskal's algorithm.
\fBKrusgen\fP can also export the maze into text (\fB\-t\fP)  or bmp (\fB\-b\fP) file, if none of the options is used, maze is exported to \fBstdout\fP as text.
.PP
Supported options:
.PP
\fB\-h\fP, \fB\-\-help\fP \- show help message
.PP
\fB\-x\fP <\fB\fCpositive, odd number\fR> \- define map width, if not used, width is set automatically to 31
.PP
\fB\-y\fP <\fB\fCpositive, odd number\fR> \- define map height, if not used, height is set automatically to 31
.PP
\fB\-f\fP, \fB\-\-frame\fP \- add borders to maze
.PP
\fB\-s\fP, \fB\-\-seed\fP <\fB\fCseed\fR> \- set random generator's seed
.PP
\fB\-v\fP, \fB\-\-version\fP \- displays krusgen version
.PP
\fB\-t\fP, \fB\-\-txt\fP <\fB\fCfilename\fR> \- export maze as text to given filename
.PP
\fB\-b\fP, \fB\-\-bmp\fP <\fB\fCfilename\fR> \- export maze as bitmap (BMP2 format) to given filename
.PP
\fB\-w\fP, \fB\-\-wall\fP <\fB\fChex color\fR> \- define wall color in bitmap, if not used wall color is white
.PP
\fB\-a\fP, \fB\-\-air\fP <\fB\fChex color\fR> \- define air color in bitmap, if not used air color is black
.PP
\fB\-W\fP, \fB\-\-wallchar\fP <\fB\fCcharacter\fR> \- define wall character, if not used wall character is #
.PP
\fB\-A\fP, \fB\-\-airchar\fP <\fB\fCcharacter\fR> \- define air character, if not used air character is ' '
.SH EXAMPLES
.PP
\fBkrusgen \-x 51 \-y 17\fP \- export maze of 51 characters width and 17 characters height to \fBstdout\fP as text.
.PP
\fBkrusgen \-b test.bmp\fP \- export maze of default dimensions as image to \fBtest.bmp\fP file.
.PP
\fBkrusgen \-t test.txt\fP \- export maze of default dimensions as text to \fBtest.txt\fP file.
.PP
\fBkrusgen \-y 25 \-x 37 \-b colors.bmp \-a FF0000 \-w FFFF00\fP \- export maze of 37 characters width and 25 characters height as image with red air and yellow walls to \fBcolors.bmp\fP file.
.PP
\fBkrusgen \-W o \-A (\fP \- export maze which has 'o' as walls and '(' as air to \fBstdout\fP\&.
.SH AUTHOR
.PP
Written by Miłosz Ryguła \[la]miloszrygula@interia.pl\[ra]\&.
\fBKrusgen\fP uses \fBlibobos\fP library created by Jacek Wieczorek \[la]mrjjot@gmail.com\[ra]\&.
.SH REPORTING BUGS
.PP
Bugs can be reported as issues on Krusgen's Github page: \[la]https://github.com/mazaczysko/krusgen/issues\[ra]
.SH COPYRIGHT
.PP
\fBKrusgen\fP and \fBlibobos\fP are both licensed under GNU GPLv3 license.
.PP
Copyright (C) 2017 Miłosz Ryguła \[la]miloszrygula@interia.pl\[ra]
Free Software Foundation, Inc.  License GPLv3+: GNU
GPL version 3 or later \[la]http://gnu.org/licenses/gpl.html\[ra]\&.
This is free software: you are free  to  change  and  redistribute  it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
krusgen (1.3-2) trusty; urgency=low

  * Minor changes.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Thu, 29 Jun 2017 00:24:00 +0200

krusgen (1.3-1) trusty; urgency=low

  * Bump to v1.3.
  * Add changing wall and air character option.
  * Minor changes.
  * Fix bugs.
  * Update help, README, manual.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sat, 13 May 2017 19:00:00 +0200

krusgen (1.2-1) trusty; urgency=low

  * Bump to v1.2.
  * Add frame option.
  * Add seed option.
  * Add version option.
  * Minor changes.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Wed, 26 Apr 2017 20:58:00 +0200

krusgen (1.1-4) trusty; urgency=low

  * Minor changes.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sun, 23 Apr 2017 22:31:00 +0200


krusgen (1.1-3) trusty; urgency=low

  * Add Travis.
  * Improve README file.
  * Fix bugs.
  * Add man page.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sun, 23 Apr 2017 22:22:00 +0200


krusgen (1.1-2) trusty; urgency=low

  * Minor changes.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sun, 23 Apr 2017 01:05:00 +0200


krusgen (1.1-1) trusty; urgency=low

  * Multiple bug fixes.
  * Security improvements.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sun, 23 Apr 2017 00:32:00 +0200


krusgen (1.0-1) trusty; urgency=low

  * Initial release.

 -- Miłosz Ryguła <miloszrygula@interia.pl>  Sat, 22 Apr 2017 20:26:00 +0200


src/krusgen.c
/*
	krusgen - a Kruskal's algorithm maze generator
	Copyright (C) 2017 Miłosz Ryguła <miloszrygula@interia.pl>

	This file is part of krusgen.

	Krusgen is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

 	Krusgen is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <inttypes.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <obos/obos.h>
#include "maze.h"
#include "krusgen.h"

#define KRUSGEN_VERSION "v1.3"

uint16_t flags = 0;

int main( int argc, char **argv )
{
	int i, badarg;
	char wallchr = '#', airchr = ' ';
	char * outputfilename = NULL;
	uint32_t colwall = 0xFFFFFF, colair = 0x000000;
	FILE *outfile = stdout;
	maze.width = 31;
	maze.height = 31;
	unsigned int seed = time( NULL );
	for( i = 1; i < argc; i++)
	{
		badarg = 1;
		if( !strcmp( argv[i], "-x" ) )
		{
			if( i + 1 >= argc || !sscanf( argv[++i], "%d", &maze.width ) )
			{
				fprintf(stderr, "%s: bad value for %s\n", argv[0], argv[i]);
				return 1;
			}
			if( mazeInit( ) != 0)
			{
				fprintf(stderr, "%s: wrong dimensions!\n", argv[0] );
				free( maze.maze );
				badarg = 1;
				return 1;
			}
			badarg = 0;
		}

		if( !strcmp( argv[i], "-y" ) )
		{
			if( i + 1 >= argc || !sscanf( argv[++i], "%d", &maze.height) )
			{
				fprintf(stderr, "%s: bad value for %s\n", argv[0], argv[i]);
				return 1;
			}
			if( mazeInit( ) != 0)
			{
				fprintf(stderr, "%s: wrong dimensions!\n", argv[0] );
				free( maze.maze );
				badarg = 1;
				return 1;
			}
			badarg = 0;
		}

		if( !strcmp( argv[i], "-h" ) || !strcmp( argv[i], "--help" ) )
		{
			printf(	"%s " KRUSGEN_VERSION " - Kruskal's algorithm maze generator\n" \
					"Usage: %s [OPTIONS]\n\n" \
					"Supported options:\n" \
					"\t-h, --help - show help\n" \
					"\t-v, --version - show Krusgen's version\n" \
					"\t-x [positive, odd number] - define map width\n" \
					"\t-y [positive, odd number] - define map height\n" \
					"\t-f, --frame - add borders to maze\n" \
					"\t-s, --seed [seed] - define seed to generate maze\n" \
					"\t-v, --version - displays krusgen version\n" \
					"\t-t, --txt [file name] - export maze to txt file\n" \
					"\t-b, --bmp [file name] - export maze to bmp file\n" \
					"\t-w, --wall [hexadecimal] - define wall color in bmp\n" \
					"\t-a, --air [hexadecimal] - define air color in bmp\n" \
					"\t-W, --wallchar [character] - define wall character\n" \
					"\t-A, --airchar [character] - define air character\n"
					, argv[0], argv[0] );
			badarg = 0;
			return 0;
		}


		if( !strcmp( argv[i], "--txt" ) || !strcmp( argv[i], "-t" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			outputfilename = argv[++i];
			flags |= FLAG_TXT;
			badarg = 0;
		}

		if( !strcmp( argv[i], "--bmp" ) || !strcmp( argv[i], "-b" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			outputfilename = argv[++i];
			flags |= FLAG_BMP;
			badarg = 0;
		}

		if( !strcmp( argv[i], "--air" ) || !strcmp( argv[i], "-a" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			if( sscanf( argv[++i], "%06x", &colair ) != 1)
			{
				fprintf( stderr, "%s: bad value '%s'\n", argv[0], argv[i] );
				return 1;
			}
			badarg = 0;
		}

		if( !strcmp( argv[i], "--wall" ) || !strcmp( argv[i], "-w" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			if( sscanf( argv[++i], "%06x", &colwall ) != 1)
			{
				fprintf( stderr, "%s: bad value '%s'\n", argv[0], argv[i] );
				return 1;
			}
			badarg = 0;
		}

		if( !strcmp( argv[i], "-f" ) || !strcmp( argv[i], "--frame" ) )
		{
			flags |= FLAG_FRAME;
			badarg = 0;
		}

		if( !strcmp( argv[i], "--seed" ) || !strcmp( argv[i], "-s" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}

			if( sscanf( argv[++i], "%u", &seed ) != 1)
			{
				fprintf( stderr, "%s: bad value '%s'\n", argv[0], argv[i] );
				return 1;
			}
			badarg = 0;
		}

		if( !strcmp( argv[i], "-v") || !strcmp( argv[i], "--version" ) )
			{
				printf("%s " KRUSGEN_VERSION "\n", argv[0] );
				badarg = 0;
				return 0;
			}

		if( !strcmp( argv[i], "--wallchar" ) || !strcmp( argv[i], "-W" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			wallchr = argv[++i][0];
			badarg = 0;
		}

		if( !strcmp( argv[i], "--airchar" ) || !strcmp( argv[i], "-A" ) )
		{
			if( i + 1 >= argc )
			{
				fprintf(stderr, "%s: missing value for '%s'\n", argv[0], argv[i] );
				return 1;
			}
			airchr =  argv[++i][0];
			badarg = 0;
		}

		if (badarg)
		{
			fprintf(stderr, "%s: bad argument '%s'\n", argv[0], argv[i]);
			return 1;
		}
	}

	if( maze.width % 2 == 0 || maze.height % 2 == 0 )
	{
		fprintf( stderr, "%s: dimensions must be odd!\n", argv[0]);
		return 1;
	}

	if( mazeInit( ) == 1 )
	{
		fprintf( stderr, "%s: memory allocation error!\n", argv[0]);
		return 1;
	}

	srand( seed );
	mazeGrid( );
	wallsList( );
	mazeGen( );

	if( flags & FLAG_BMP && flags & FLAG_TXT )
	{
		fprintf( stderr, "%s: cannot export to .bmp and .txt files at once!\n", argv[0]);
		return 1;
	}

	if( outputfilename != NULL)
	{
		outfile = fopen( outputfilename, "w");

		if( outfile == NULL )
		{
			fprintf(stderr, "%s: cannot open file!\n", argv[0] );
			return 1;
		}
	}

	if( flags & FLAG_BMP )
		mazeBmp( outfile, colwall, colair );
	else
		mazeDraw( outfile, wallchr, airchr );

	fclose( outfile );
	return 0;
}
